src/SessionManager/ProgramSettings.cc

```diff
@@ -0,0 +1,126 @@
+
+#include <cxxopts.hpp>
+
+#ifdef _BOOST_FILESYSTEM_
+#include <boost/filesystem.hpp>
+namespace fs = boost::filesystem;
+#else
+#include <filesystem>
+
+namespace fs = std::filesystem;
+#endif
+
+using namespace std;
+
+namespace carta {
+
+template <class T>
+void applyOptionalArgument(T& val, const string& argument_name, const cxxopts::ParseResult& results) {
+    if (results.count(argument_name)) {
+        val = results[argument_name].as<T>();
+    }
+}
+
+ProgramSettings::ProgramSettings(int argc, char** argv) {
+    cxxopts::Options options("CARTA", "CARTA Backend");
+    // clang-format off
+    // clang-format doesn't like this chain of calls
+    options.add_options()("h,help", "Print usage")
+        ("v,version", "Print version")
+        ("verbosity",
+         "Display verbose logging from level (0: off, 1: critical, 2: error, 3: warning, 4: info, 5: debug, 6: trace)",
+         cxxopts::value<int>()->default_value(to_string(verbosity)), "<level>")
+        ("no_log", "Do not output to a log file", cxxopts::value<bool>())
+        ("no_http", "Disable CARTA frontend HTTP server", cxxopts::value<bool>())
+        ("no_browser", "Prevent the frontend from automatically opening in the default browser on startup", cxxopts::value<bool>())
+        ("host", "Only listen on the specified interface (IP address or hostname)", cxxopts::value<string>()->default_value(host), "<interface>")
+        ("p,port", "Manually set port on which to host frontend files and accept WebSocket connections", cxxopts::value<int>(), "<port number>")
+        ("g,grpc_port", "Set grpc server port", cxxopts::value<int>(), "<port number>")
+        ("t,threads", "Manually set OpenMP thread pool count", cxxopts::value<int>(), "<thread count>")
+        ("top_level_folder", "Set top-level folder for data files. Files outside of this directory will not be accessible", cxxopts::value<string>(), "<path>")
+        ("frontend_folder", "Set folder to serve frontend files from", cxxopts::value<string>(), "<path>")
+        ("exit_after", "Number of seconds to stay alive after last sessions exists", cxxopts::value<int>(), "<duration>")
+        ("init_exit_after", "Number of seconds to stay alive at start if no clients connect", cxxopts::value<int>(), "<duration>")
+        ("files", "Files to load", cxxopts::value<vector<string>>(files));
+
+    options.add_options("deprecated and debug")
+        ("debug_no_auth", "Accept all incoming WebSocket connections (insecure, use with caution!)", cxxopts::value<bool>())
+        ("base", "[Deprecated] Set starting folder for data files", cxxopts::value<string>(), "<path>")
+        ("root", "[Deprecated] Use 'top_level_folder' instead", cxxopts::value<string>(), "<path>");
+    // clang-format on
+
+    options.positional_help("<file(s) or folder to open>");
+    options.parse_positional("files");
+    auto result = options.parse(argc, argv);
+
+    if (result.count("version")) {
+        cout << VERSION_ID << endl;
```

Comment:
Use spdlog instead?

Reply:
This and the help message are the two places where I suspect we shouldn't use spdlog. They're called before the logger is initialised, so they'll just end up in stdout anyway. Also, version and help text generally follows a convention, and adding logging prefixes would clutter that.

Reply:
> This and the help message are the two places where I suspect we shouldn't use spdlog. They're called before the logger is initialised, so they'll just end up in stdout anyway. Also, version and help text generally follows a convention, and adding logging prefixes would clutter that.

Good point!